In [1]:
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt

### PLAN

1. Pobranie danych
2. Przejrzeć co mamy
    - wielkosc data setu
    - jakie kolumny
    - jakiego typu dane
    - podejrzec probke
    - jesli dane sa we wlasciwym formacie: proste statystyki
3. Konwersja danych na odpowiedni format
4. Co z brakujacymi wartosciami?

#### Zaimportuj i przejrzyj dane

In [2]:
dane = pd.read_csv('data/DataWorkshop.csv')

In [3]:
print('Rozmiar danych: ', dane.shape)

Rozmiar danych:  (229929, 30)


In [4]:
print('Kolumny w datasecie: ')
dane.columns

Kolumny w datasecie: 


Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków'],
      dtype='object')

In [5]:
print('Typy danych: ')
dane.dtypes

Typy danych: 


Data zdarzenia                        object
Godzina                               object
GPS x                                 object
GPS y                                 object
Miejscowość                           object
Ulica                                 object
Numer domu                            object
Numer drogi                           object
KM HM (Pikietaż)                      object
Obszar                                object
Skrzyżowanie                          object
Geometria drogi                       object
Charakterystyka miejsca zdarzenia     object
Oświetlenie                           object
Warunki atmosferyczne                 object
Rodzaj zdarzenia                      object
Wina                                  object
Rodzaj drogi                          object
Nawierzchnia drogi                    object
Stan nawierzchni                      object
Sygnalizacja świetlna                 object
Oznakowanie poziome                   object
Liczba pas

In [6]:
print('Brakujace wartosci: ')

missing = dane.isnull().sum()
print(missing)

Brakujace wartosci: 
Data zdarzenia                            0
Godzina                                   0
GPS x                                     0
GPS y                                     0
Miejscowość                               0
Ulica                                 20728
Numer domu                            93884
Numer drogi                               0
KM HM (Pikietaż)                     194888
Obszar                                    1
Skrzyżowanie                         166779
Geometria drogi                       66008
Charakterystyka miejsca zdarzenia         2
Oświetlenie                               0
Warunki atmosferyczne                     1
Rodzaj zdarzenia                          0
Wina                                      0
Rodzaj drogi                              1
Nawierzchnia drogi                        1
Stan nawierzchni                          1
Sygnalizacja świetlna                     0
Oznakowanie poziome                       2
Liczba pasó

In [7]:
print('Brakujace wartosci jako procent zbioru, %: ')
index=[]
lst=[]
for item in missing.iteritems():
    index.append(item[0])
    lst.append(round((item[1]/dane.shape[0])*100,4))
missing_percentage=pd.Series(lst, index=index)
print(missing_percentage)

Brakujace wartosci jako procent zbioru, %: 
Data zdarzenia                        0.0000
Godzina                               0.0000
GPS x                                 0.0000
GPS y                                 0.0000
Miejscowość                           0.0000
Ulica                                 9.0150
Numer domu                           40.8317
Numer drogi                           0.0000
KM HM (Pikietaż)                     84.7601
Obszar                                0.0004
Skrzyżowanie                         72.5350
Geometria drogi                      28.7080
Charakterystyka miejsca zdarzenia     0.0009
Oświetlenie                           0.0000
Warunki atmosferyczne                 0.0004
Rodzaj zdarzenia                      0.0000
Wina                                  0.0000
Rodzaj drogi                          0.0004
Nawierzchnia drogi                    0.0004
Stan nawierzchni                      0.0004
Sygnalizacja świetlna                 0.0000
Oznakowanie

In [8]:
# Probka danych
dane.sample(50)

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
22537      2016-06-14  15:00:00  19*07'338  50*49'071           CZĘSTOCHOWA   
201198     2019-06-03  08:49:00  19*20'409  50*55'381              KŁOMNICE   
99132      2017-10-25  22:01:00  19*00'166  50*49'059             GORZELNIA   
65539      2017-03-26  17:00:00  19*18'432  50*10'061              JAWORZNO   
4835       2016-02-10  14:10:00  19*16'501  50*21'511      DĄBROWA GÓRNICZA   
119264     2018-02-26  17:10:00  19*02'464  50*15'456              KATOWICE   
226630     2019-12-13  09:50:00  19*17'592  50*16'518               SŁAWKÓW   
156093     2018-10-20  19:15:00  18*37'544  49*49'506      KOŃCZYCE WIELKIE   
155526     2018-10-17  11:00:00  19*02'580  49*49'383         BIELSKO-BIAŁA   
129366     2018-05-04  19:05:00  18*27'030  50*23'443                  ŁANY   
57181      2017-01-29  17:45:00  18*56'186  49*33'029              KONIAKÓW   
209753     2019-09-10  16:05:00  18*51'508  50*19'131           RUDA ŚLĄSKA   
46858      2016-11-26  11:59:00  19*07'025  50*48'426           CZĘSTOCHOWA   
47321      2016-11-28  16:45:00  19*13'032  50*19'324      DĄBROWA GÓRNICZA   
93793      2017-09-23  11:55:00  19*11'235  50*18'123             SOSNOWIEC   
44251      2016-10-31  15:00:00  19*03'441  50*16'092              KATOWICE   
19300      2016-05-24  09:10:00  18*54'479  50*28'451    MIASTECZKO ŚLĄSKIE   
169465     2019-01-06  17:00:00  19*03'288  50*49'536           CZĘSTOCHOWA   
118919     2018-02-24  15:30:00  18*47'332  49*49'181               WIŚLICA   
118692     2018-02-20  11:30:00  19*16'066  50*37'019        ŻARKI-LETNISKO   
204253     2019-08-07  10:00:00  19*01'080  50*16'343              KATOWICE   
52656      2016-12-28  15:00:00  18*59'285  50*07'555                 TYCHY   
65253      2017-03-27  20:30:00  18*54'040  50*26'231         NAKŁO ŚLĄSKIE   
167943     2018-12-27  17:35:00  18*58'565  50*14'335              KATOWICE   
147324     2018-08-30  20:55:00  18*50'196  50*15'415           RUDA ŚLĄSKA   
118222     2018-02-19  09:30:00  18*40'197  50*17'125               GLIWICE   
51670      2016-12-21  21:28:00  18*39'507  50*17'097               GLIWICE   
150233     2018-09-17  10:05:00  19*00'562  50*19'474  SIEMIANOWICE ŚLĄSKIE   
26386      2016-07-13  18:15:00  18*59'068  50*07'362                 TYCHY   
125027     2018-04-06  14:00:00  19*09'328  50*13'453             SOSNOWIEC   
126623     2018-04-16  12:40:00  19*00'250  50*16'186              KATOWICE   
64694      2017-03-22  19:01:00  18*24'092  50*02'265                 PSZÓW   
114882     2018-01-27  05:49:00  19*14'490  49*40'123                ŻYWIEC   
84639      2017-08-01  09:15:00  19*00'378  50*17'070              KATOWICE   
218925     2019-10-30  14:15:00  18*41'109  50*00'472                  ŻORY   
136729     2018-06-20  19:00:00  19*07'230  50*40'193               ROMANÓW   
179732     2019-03-10  05:45:00  18*56'455  49*58'436              PSZCZYNA   
123309     2018-03-25  10:20:00  19*24'171  50*57'405                WIDZÓW   
138974     2018-07-06  10:35:00  18*55'259  50*09'344               MIKOŁÓW   
5511       2016-02-13  10:00:00  18*49'161  50*19'103                ZABRZE   
11446      2016-03-30  16:00:00  18*56'388  50*22'187       PIEKARY ŚLĄSKIE   
226584     2019-12-13  06:30:00  19*00'575  50*09'543                 TYCHY   
46062      2016-11-19  16:34:00  19*07'359  50*47'042           CZĘSTOCHOWA   
143559     2018-08-05  13:35:00  18*40'254  50*17'335               GLIWICE   
133022     2018-05-27  10:30:00  19*06'167  50*48'437           CZĘSTOCHOWA   
227200     2019-12-16  15:00:00  19*03'471  49*50'452         BIELSKO-BIAŁA   
17746      2016-05-13  12:30:00  18*49'317  50*12'289               MIKOŁÓW   
211100     2019-09-18  13:48:00  18*56'248  50*54'223               KŁOBUCK   
216079     2019-10-15  13:00:00  18*41'223  50*01'435                  ŻORY   
201190     2019-07-18  09:00:00  18*55'014  50*20'327    

### Konwersja typu danych

#### Data i czas

In [9]:
# Utworzenie kopii zbioru danych. Dalsza konwersja odbywac sie bedzie na kopii danych
df=dane.copy()

In [10]:
# Data zdarzenia i Godzina scalone do jednej kolumny
print('kolumna DATA ZDARZENIA: ', df['Data zdarzenia'].dtypes)
print('kolumna GODZINA: ', df['Data zdarzenia'].dtypes)
df['data_full'] = df['Data zdarzenia'] + ' ' + df['Godzina']
df.sample(10)

kolumna DATA ZDARZENIA:  object
kolumna GODZINA:  object


Data zdarzenia   Godzina      GPS x      GPS y       Miejscowość  \
124419     2018-04-01  02:00:00  19*08'026  50*17'475         SOSNOWIEC   
162345     2018-11-27  16:45:00  18*59'507  50*14'579          KATOWICE   
13950      2016-04-18  15:10:00  19*02'029  50*07'013             TYCHY   
160099     2018-11-13  20:25:00  19*13'337  50*19'517  DĄBROWA GÓRNICZA   
216888     2019-10-18  22:10:00  19*06'073  50*18'243           CZELADŹ   
25976      2016-07-08  13:00:00  18*57'043  50*17'345           CHORZÓW   
33585      2016-09-03  10:55:00  19*04'549  50*28'002         PYRZOWICE   
74530      2017-05-27  16:30:00  18*55'083  50*19'027    ŚWIĘTOCHŁOWICE   
111844     2018-01-09  10:56:00  19*07'281  50*47'450       CZĘSTOCHOWA   
162809     2018-11-30  16:15:00  18*40'104  50*40'146         LUBLINIEC   

                           Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
124419                 BĘDZIŃSKA         25      P7620S              NaN   
162345               KOCHŁOWICKA        NaN      P6479S              NaN   
13950                OŚWIĘCIMSKA        142          44               37   
160099               KOSMONAUTÓW        NaN   G200 077S              NaN   
216888                   NORWIDA        17A          I2              NaN   
25976   GEN.HENRYKA DĄBROWSKIEGO         34    G210079S              NaN   
33585               TRANSPORTOWA        NaN        W913              NaN   
74530                ŁAGIEWNICKA        NaN      P9807S              NaN   
111844      ALEJA NIEPODLEGŁOŚCI        NaN      P6614S              NaN   
162809                 LISOWICKA         19        W906              NaN   

                      Obszar  ... Oznakowanie poziome  \
124419     Obszar zabudowany  ...                Jest   
162345  Obszar niezabudowany  ...                Jest   
13950      Obszar zabudowany  ...                Jest   
160099     Obszar zabudowany  ...                Jest   
216888     Obszar zabudowany  ...                Jest   
25976      Obszar zabudowany  ...                Jest   
33585   Obszar niezabudowany  ...                Jest   
74530      Obszar zabudowany  ...              Nie ma   
111844     Obszar zabudowany  ...                Jest   
162809     Obszar zabudowany  ...                Jest   

       Liczba pasów w zdarzeniu  \
124419                      2.0   
162345                      1.0   
13950                       NaN   
160099                      2.0   
216888                      2.0   
25976                       3.0   
33585                       2.0   
74530                       NaN   
111844                      4.0   
162809                      2.0   

                                  Zachowanie kierującego Zachowanie pieszego  \
124419                             Nieprawidłowe cofanie                 NaN   
162345       Niezachowanie bezp. odleg. między pojazdami                 NaN   
13950                           Nieprawidłowe: wymijanie                 NaN   
160099                                    Inne przyczyny                 NaN   
216888                                    Inne przyczyny                 NaN   
25976                             Nieprawidłowe omijanie                 NaN   
33585        Niezachowanie bezp. odleg. między pojazdami                 NaN   
74530   Nieprawidłowe wymijanie; Nieprawidłowe wymijanie                 NaN   
111844                                               NaN                 NaN   
162809       Niezachowanie bezp. odleg. między pojazdami                 NaN   

       Inne przyczyny Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu  \
124419            NaN                          0                           0   
162345            NaN                          0                           0   
13950             NaN                          0                           0   
160099            NaN                          0                           0   
216888            NaN                          0 

In [11]:
# Data zdarzenia i Godzina do typu datetime
df['data_full'] = pd.to_datetime(df['data_full'], format='%Y-%m-%d %H:%M:%S')
df.sample(10)
# TO-DO: Jaki to dzien tygodnia? Jaki to dzien miesiaca? To dzien czy noc? Jaka faza ksiezyca itd..

Data zdarzenia   Godzina      GPS x      GPS y      Miejscowość  \
95902      2017-10-07  12:15:00  18*56'281  50*22'359  PIEKARY ŚLĄSKIE   
71332      2017-05-08  07:10:00  18*48'273  50*26'093  TARNOWSKIE GÓRY   
97855      2017-10-17  08:50:00  19*07'271  50*16'462        SOSNOWIEC   
63962      2017-03-18  00:10:00  18*32'027  50*05'590           RYBNIK   
9666       2016-03-11  02:00:00  18*32'201  50*05'382           RYBNIK   
217972     2019-10-25  08:20:00  18*57'364  50*10'063         KATOWICE   
9928       2016-03-14  10:50:00  18*56'164  50*20'358            BYTOM   
174531     2019-02-04  14:20:00  19*17'352  50*32'577          MYSZKÓW   
158909     2018-11-05  12:26:00  18*34'250  50*05'548           RYBNIK   
93395      2017-09-24  18:00:00  19*05'305  50*16'562         KATOWICE   

                           Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
95902                   OGRODOWA         16    G280146S              NaN   
71332          JANA ŚNIADECKIEGO          1   G270 298S              NaN   
97855                KILIŃSKIEGO         25    G120022S              NaN   
63962               JANA KOTUCZA        NaN          78              NaN   
9666                     HALLERA        12A          I2              NaN   
217972                 HORTENSJI         19    G100240S              NaN   
9928         ALEJA JANA PAWŁA II        NaN        W911              NaN   
174531                PUŁASKIEGO        70A        W793              NaN   
158909               PRZEMYSŁOWA         26      P7013S              NaN   
93395   WALENTEGO ROŹDZIEŃSKIEGO        201         S86              NaN   

                      Obszar  ... Oznakowanie poziome  \
95902      Obszar zabudowany  ...              Nie ma   
71332      Obszar zabudowany  ...              Nie ma   
97855      Obszar zabudowany  ...                Jest   
63962      Obszar zabudowany  ...                Jest   
9666       Obszar zabudowany  ...              Nie ma   
217972     Obszar zabudowany  ...                Jest   
9928       Obszar zabudowany  ...                Jest   
174531     Obszar zabudowany  ...                Jest   
158909     Obszar zabudowany  ...                Jest   
93395   Obszar niezabudowany  ...                Jest   

       Liczba pasów w zdarzeniu  \
95902                       NaN   
71332                       2.0   
97855                       NaN   
63962                       NaN   
9666                        NaN   
217972                      1.0   
9928                        NaN   
174531                      2.0   
158909                      2.0   
93395                       NaN   

                                   Zachowanie kierującego  \
95902                                                 NaN   
71332                                                 NaN   
97855                              Nieprawidłowe omijanie   
63962                                                 NaN   
9666                Nieudzielenie pierwszeństwa przejazdu   
217972                            Nieprawidłowe wymijanie   
9928            Niezachowanie bezp. odl. między pojazdami   
174531        Niedostosowanie prędkości do warunków ruchu   
158909  Nieprawidłowe omijanie; Nieprzestrzeganie znak...   
93395                                                 NaN   

                                      Zachowanie pieszego Inne przyczyny  \
95902   Nieostrożne wejście na jezdnię: przed jadącym ...            NaN   
71332                                                 NaN    Stan jezdni   
97855                                                 NaN            NaN   
63962                                                 NaN    Nieustalone   
9666                                                  NaN            NaN   
217972                                                NaN            NaN   
9928                                                  NaN            NaN   
174531                                                NaN    

#### Wspolrzedne geograficzne

Domyslny zapis wspolrzednych to D\*M'S([D]egree, [M]inute, [S]cond) np. 18\*53'289. Jednak, gdy pierwsza cyfra sekund jest zero, czasem wystepuje zapis D\*M''S np. 18\*53''11. Wystepowaly takze bledne zapisy takie jak 4\*\*56'569 i 50\*09'\*01.

In [12]:
def dmsToDecimal(coordinate_as_string):
    try:
        # Zapis domyslny to D*M'S([D]egree, [M]inute, [S]cond) zeby sprawnie podzielic kazda wspolrzedna na liste trzyelementowa
        # zamieniam ' na *, a nastepnie dziele orzymany rezultat znakami *. W trakcie konwersji danych napotkalam rozne nieprawidlowe
        # zapisy opisane w komorce powyzej. Wszystkie wystepujace nietypowe znaki takie jak **, '* i '' zamieniam na * i dziele znakiem *
        degree, minute, second = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
    except:
        # Jesli nie dzieli na liste trzyelementowa to najprawdopodobniej trzeci element jest pusty(lista dwuelementowa).
        # Wtedy pierwszy element to stopien, drugi minuta. Sekundy zas sa rowne 0.
        degree, minute = str(coordinate_as_string).replace("**", "*").replace("'*", "*").replace("''", "*").replace("'", '*').split("*")
        second=0
    # W niektorych wypadkach udaje sie podzielic wsadowego stringa tak aby uzyskac ostatnia czesc jako pusty string.
    #Ponizej przeksztalcam pustego stringa w wartosc 0
    if second == "" or second == " ":
        second=0
    # Wlasciwe przeksztalcenie wspolrzednej z wartosci katowych na wartosc liczbowa
    coordinate_as_float = float(degree) + float(minute)/60 + float(second)/(60*60)
    return coordinate_as_float
# Przygotowana wyzej funkcje wywoluje na calych kolumnach dancych
df['GPS x'] = df['GPS x'].apply(lambda x: dmsToDecimal(x))
df['GPS y'] = df['GPS y'].apply(lambda x: dmsToDecimal(x))
df.sample(10)

Data zdarzenia   Godzina      GPS x      GPS y Miejscowość  \
219991     2019-11-05  15:55:00  18.680000  50.109167      RYBNIK   
215552     2019-10-12  11:45:00  19.445556  50.775000       JANÓW   
204769     2019-08-10  09:30:00  18.780000  50.340833     GLIWICE   
198871     2019-07-03  03:19:00  19.962778  50.569444   ŻARNOWIEC   
103924     2017-11-22  16:50:00  19.170278  50.375278      BĘDZIN   
42973      2016-10-01  03:30:00  19.155278  50.376944     CZELADŹ   
92472      2017-09-20  08:30:00  19.189167  50.328611   MYSŁOWICE   
9911       2016-03-17  07:45:00  18.733611  50.382222     GLIWICE   
85610      2017-08-08  15:30:00  18.792500  50.410000      ZABRZE   
24779      2016-06-30  10:50:00  19.345278  50.533333   PODWARPIE   

                     Ulica Numer domu Numer drogi KM HM (Pikietaż)  \
219991          MIKOŁOWSKA   NIEZNANE        W925              NaN   
215552       CZĘSTOCHOWSKA          1          46              NaN   
204769                 NaN        NaN          A4            313,5   
198871          MIECHOWSKA          3      P1769S              NaN   
103924                 NaN        NaN          86             14,8   
42973                  NaN        NaN        K94B             18,3   
92472                  NaN        NaN          S1            548,4   
9911          JAGIELLOŃSKA        NaN      P7216S              NaN   
85610   KS.JÓZEFA LONDZINA          2    G145421S              NaN   
24779                  NaN        NaN           1            521,1   

                      Obszar  ... Oznakowanie poziome  \
219991     Obszar zabudowany  ...                Jest   
215552     Obszar zabudowany  ...                Jest   
204769  Obszar niezabudowany  ...                Jest   
198871     Obszar zabudowany  ...              Nie ma   
103924  Obszar niezabudowany  ...                Jest   
42973   Obszar niezabudowany  ...                Jest   
92472   Obszar niezabudowany  ...                Jest   
9911       Obszar zabudowany  ...                Jest   
85610      Obszar zabudowany  ...              Nie ma   
24779   Obszar niezabudowany  ...                Jest   

       Liczba pasów w zdarzeniu                       Zachowanie kierującego  \
219991                      2.0        Nieustąpienie pierwszeństwa przejazdu   
215552                      2.0        Nieustąpienie pierwszeństwa przejazdu   
204769                      4.0  Niezachowanie bezp. odleg. między pojazdami   
198871                      2.0  Niedostosowanie prędkości do warunków ruchu   
103924                      NaN  Niezachowanie bezp. odleg. między pojazdami   
42973                       NaN                                          NaN   
92472                       NaN          Nieprawidłowe zmienianie pasa ruchu   
9911                        2.0                      Nieprawidłowe: omijanie   
85610                       NaN                       Nieprawidłowe omijanie   
24779                       NaN  Niezachowanie bezp. odleg. między pojazdami   

       Zachowanie pieszego                Inne przyczyny  \
219991                 NaN                           NaN   
215552                 NaN                           NaN   
204769                 NaN                           NaN   
198871                 NaN                           NaN   
103924                 NaN                           NaN   
42973                  NaN  Obiekty, zwierzęta na drodze   
92472                  NaN                           NaN   
9911                   NaN                           NaN   
85610                  NaN                           NaN   
24779                  NaN                           NaN   

       Liczba rannych w zdarzeniu Liczba zabitych w zdarzeniu Liczba kolizji  \
219991                          0                           0              1   
215552                          0                           0              1   
204769                          0                           0     